File: demo/generate_data.py
#!/usr/bin/env python3
"""Generate the synthetic OHLC data used by the wichtelm-app demo backtest.

The output is fully deterministic, so re-running this script reproduces the
exact CSV files committed under demo/data/.

The price path is built from:

  * a slow trend regime (half a sine wave over the whole window): the market
    rises for the first ~60 days and falls for the last ~60, so the daily
    trend filter admits long entries early and short entries late;
  * three oscillations with incommensurate periods. They beat against each
    other, so the swing the strategy trades into varies in size from one
    cycle to the next - which is what gives the backtest a realistic mix of
    winning and losing trades rather than a perfectly periodic outcome;
  * a small irregular wobble so bars are not perfectly smooth.

The 1d series is aggregated from the 1h series, so the two timeframes are
always mutually consistent (the daily bar spans its 24 hourly bars).
"""

import math
from datetime import datetime, timedelta, timezone

DAYS = 120
HOURS = DAYS * 24
START = datetime(2024, 1, 1, tzinfo=timezone.utc)

BASE = 118.0
TREND_AMP = 52.0
TREND_PERIOD_DAYS = 240.0   # half a cycle over the 120-day window


def close_price(i: int) -> float:
    """Hourly close for hour index i."""
    day = i / 24.0
    trend = TREND_AMP * math.sin(2.0 * math.pi * day / TREND_PERIOD_DAYS)
    osc = (7.0 * math.sin(2.0 * math.pi * i / 54.0)
           + 4.0 * math.sin(2.0 * math.pi * i / 89.0 + 0.7)
           + 2.5 * math.sin(2.0 * math.pi * i / 31.0 + 1.9))
    wobble = 1.1 * math.sin(i * 0.37) + 0.7 * math.sin(i * 0.13)
    return BASE + trend + osc + wobble


def hourly_bars():
    bars = []
    prev_close = close_price(0)
    for i in range(HOURS):
        close = close_price(i)
        open_ = prev_close if i > 0 else close
        rng = 0.5 + 0.35 * abs(math.sin(i * 0.9))
        high = max(open_, close) + rng
        low = min(open_, close) - rng
        bars.append({
            "time": START + timedelta(hours=i),
            "open": open_,
            "high": high,
            "low": low,
            "close": close,
        })
        prev_close = close
    return bars


def daily_bars(hourly):
    bars = []
    for d in range(DAYS):
        window = hourly[d * 24:(d + 1) * 24]
        bars.append({
            "time": START + timedelta(days=d),
            "open": window[0]["open"],
            "high": max(b["high"] for b in window),
            "low": min(b["low"] for b in window),
            "close": window[-1]["close"],
        })
    return bars


def iso(dt: datetime) -> str:
    return dt.strftime("%Y-%m-%dT%H:%M:%SZ")


def write_csv(path: str, bars):
    with open(path, "w", encoding="utf-8") as handle:
        handle.write("time,open,high,low,close\n")
        for bar in bars:
            handle.write("{},{:.2f},{:.2f},{:.2f},{:.2f}\n".format(
                iso(bar["time"]), bar["open"], bar["high"], bar["low"], bar["close"]))


def main():
    hourly = hourly_bars()
    daily = daily_bars(hourly)
    write_csv("demo/data/DEMO_1h.csv", hourly)
    write_csv("demo/data/DEMO_1d.csv", daily)
    print(f"wrote demo/data/DEMO_1h.csv ({len(hourly)} bars)")
    print(f"wrote demo/data/DEMO_1d.csv ({len(daily)} bars)")


if __name__ == "__main__":
    main()

File: demo/GenerateData.java
// Generates the synthetic OHLC data for the wichtelm-app demos.
//
// Single-file Java program — no build step, JDK-only. Run from the repo root:
//
//   java demo/GenerateData.java
//
// Output is fully deterministic (no RNG; everything is closed-form), so
// re-running reproduces the committed demo/data/TSTX_*.csv exactly.
//
// One synthetic instrument, "TSTX" (a deliberately fake name per CLAUDE.md §17 —
// never a real ticker), is emitted at four timeframes: 1h, 4h, 1d, 1w. The 4h /
// 1d / 1w series are aggregated from the 1h base, so every timeframe is mutually
// consistent. The window is 156 whole weeks (1092 days ≈ 3 years) starting on a
// Monday, so weekly bars align to week boundaries.
//
// The price path layers, so the SAME data yields tradeable structure at every
// timeframe (no perfectly periodic outcomes):
//   * a ~2-year regime cycle (bull → bear → bull) — the long trend filters key on;
//   * ~26-day and ~41-day swings — what the daily/weekly strategies trade;
//   * a ~5-day swing and an intraday cycle — for the 4h / 1h strategies;
//   * a small irregular wobble so bars are not perfectly smooth.
// Volume is synthetic but always positive (the volume-gated demo needs it).
//
// The data is continuous (a bar every period, weekends included) — that is what
// makes the charts clean; it is explicitly synthetic, not a model of real hours.

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public final class GenerateData {

    static final int WEEKS = 156;                 // 156 whole weeks
    static final int DAYS = WEEKS * 7;            // 1092 days ≈ 3 years
    static final int HOURS = DAYS * 24;           // 26208 hours
    static final Instant START = Instant.parse("2021-01-04T00:00:00Z"); // a Monday
    static final double BASE = 140.0;

    record Bar(Instant time, double open, double high, double low, double close, long volume) {}

    /** Hourly close for hour index i. */
    static double closePrice(int i) {
        double day = i / 24.0;
        double trend   = 45.0 * Math.sin(2 * Math.PI * day / 730.0);            // ~2yr: bull→bear→bull
        double swingA  = 14.0 * Math.sin(2 * Math.PI * day / 26.0);             // ~26-day swing
        double swingB  = 8.0  * Math.sin(2 * Math.PI * day / 41.0 + 0.8);       // ~41-day swing
        double daily   = 4.0  * Math.sin(2 * Math.PI * day / 5.3 + 1.1);        // ~5-day swing
        double intra   = 2.2  * Math.sin(2 * Math.PI * i / 24.0)
                       + 1.3  * Math.sin(2 * Math.PI * i / 8.0 + 0.5);          // intraday structure
        double wobble  = 0.8  * Math.sin(i * 0.37) + 0.5 * Math.sin(i * 0.131 + 2.0);
        return BASE + trend + swingA + swingB + daily + intra + wobble;
    }

    /** Always-positive synthetic hourly volume. */
    static long volume(int i) {
        double v = 40_000_000.0
                + 25_000_000.0 * Math.abs(Math.sin(2 * Math.PI * i / 37.0))
                + 10_000_000.0 * Math.abs(Math.sin(i * 0.07 + 1.0));
        return (long) v;
    }

    static List<Bar> hourlyBars() {
        List<Bar> bars = new ArrayList<>(HOURS);
        double prevClose = closePrice(0);
        for (int i = 0; i < HOURS; i++) {
            double close = closePrice(i);
            double open = i > 0 ? prevClose : close;
            double rng = close * (0.0035 + 0.0025 * Math.abs(Math.sin(i * 0.9)));
            double high = Math.max(open, close) + rng;
            double low = Math.min(open, close) - rng;
            bars.add(new Bar(START.plus(Duration.ofHours(i)), open, high, low, close, volume(i)));
            prevClose = close;
        }
        return bars;
    }

    /** Aggregate the 1h base into {@code spanHours}-wide bars (4h / 1d / 1w). */
    static List<Bar> aggregate(List<Bar> hourly, int spanHours) {
        List<Bar> out = new ArrayList<>();
        for (int s = 0; s + spanHours <= hourly.size(); s += spanHours) {
            List<Bar> w = hourly.subList(s, s + spanHours);
            double high = Double.NEGATIVE_INFINITY, low = Double.POSITIVE_INFINITY;
            long vol = 0;
            for (Bar b : w) {
                high = Math.max(high, b.high());
                low = Math.min(low, b.low());
                vol += b.volume();
            }
            out.add(new Bar(w.get(0).time(), w.get(0).open(), high, low,
                    w.get(w.size() - 1).close(), vol));
        }
        return out;
    }

    static void writeCsv(String path, List<Bar> bars) throws IOException {
        StringBuilder sb = new StringBuilder("time,open,high,low,close,volume\n");
        for (Bar b : bars) {
            sb.append(String.format(Locale.ROOT, "%s,%.2f,%.2f,%.2f,%.2f,%d\n",
                    b.time(), b.open(), b.high(), b.low(), b.close(), b.volume()));
        }
        Files.writeString(Path.of(path), sb.toString());
    }

    public static void main(String[] args) throws IOException {
        List<Bar> h1 = hourlyBars();
        List<Bar> h4 = aggregate(h1, 4);
        List<Bar> d1 = aggregate(h1, 24);
        List<Bar> w1 = aggregate(h1, 168);
        writeCsv("demo/data/TSTX_1h.csv", h1);
        writeCsv("demo/data/TSTX_4h.csv", h4);
        writeCsv("demo/data/TSTX_1d.csv", d1);
        writeCsv("demo/data/TSTX_1w.csv", w1);
        System.out.printf(Locale.ROOT, "TSTX: 1h=%d 4h=%d 1d=%d 1w=%d%n",
                h1.size(), h4.size(), d1.size(), w1.size());
    }
}
